### PR TITLE
Reconnect on cert change

### PR DESCRIPTION
**Pull request by mlohvynenko**
**Depends on**:  https://github.com/aosedge/aos_core_lib_cpp/pull/159

### CODE REVIEW

File: src/communication/tlschannel.cpp [hunk 1]
```diff
@@ -50,6 +50,8 @@ Error TLSChannel::SetTLSConfig(const String& certType)
 {
     LOG_DBG() << "Set TLS config: name=" << mName << ", certType=" << certType;
 
+    Cleanup();
```
Comment (al1img) [inline src/communication/tlschannel.cpp:53]: As I remember there were some issues cleaning up not setup TLS config. So, cleanup only if mCertType not empty and clear it in cleanup for example.

File: src/iamclient/iamclient.cpp [hunk 2]
```diff
@@ -139,6 +145,14 @@ Error IAMClient::OnNodeStatusChanged(const String& nodeID, const NodeStatus& sta
     return ErrorEnum::eNone;
 }
 
+void IAMClient::OnCertChanged(const iam::certhandler::CertInfo& info)
+{
+    LOG_DBG() << "Received on cert changed event: " << info;
```
Comment (al1img) [inline src/iamclient/iamclient.cpp:150]: `Cert changed` or `Cert changed event received`.

Should be guarded with mutex.

File: src/iamclient/iamclient.cpp [hunk 3]
```diff
@@ -181,6 +195,12 @@ Error IAMClient::ReleaseChannel()
         err = AOS_ERROR_WRAP(deleteErr);
     }
 
+#ifndef CONFIG_ZTEST
+    if (auto err = mCertHandler->UnsubscribeCertChanged(*this); !err.IsNone()) {
+        err = Error(AOS_ERROR_WRAP(err), "Can't unsubscribe from cert changed event");
```
Comment (al1img) [inline src/iamclient/iamclient.cpp:200]: AOS_ERROR_WRAP(Error(....

error message from small letter

File: src/iamclient/iamclient.cpp [hunk 4]
```diff
@@ -211,6 +231,10 @@ Error IAMClient::SetupChannel()
         mCurrentPort = cSecurePort;
 
 #ifndef CONFIG_ZTEST
+        if (auto err = mCertHandler->SubscribeCertChanged(cIAMCertType, *this); !err.IsNone()) {
```
Comment (al1img) [inline src/iamclient/iamclient.cpp:234]: ditto

File: src/smclient/smclient.cpp [hunk 5]
```diff
@@ -305,6 +305,14 @@ void SMClient::OnClockUnsynced()
     mCondVar.NotifyOne();
 }
 
+void SMClient::OnCertChanged(const iam::certhandler::CertInfo& info)
+{
+    LOG_DBG() << "Received on cert changed event: " << info;
```
Comment (al1img) [inline src/smclient/smclient.cpp:310]: `Cert changed` or `Cert changed event received`.

Should be guarded with mutex.

File: src/smclient/smclient.cpp [hunk 6]
```diff
@@ -592,6 +600,10 @@ Error SMClient::SetupChannel()
     }
 
 #ifndef CONFIG_ZTEST
+    if (err = mCertHandler->SubscribeCertChanged(cSMCertType, *this); !err.IsNone()) {
+        return Error(AOS_ERROR_WRAP(err), "failed to subscribe on cert changed event");
```
Comment (al1img) [inline src/smclient/smclient.cpp:604]: Wrap error

File: src/smclient/smclient.cpp [hunk 7]
```diff
@@ -632,6 +646,12 @@ Error SMClient::ReleaseChannel()
         return AOS_ERROR_WRAP(err);
     }
 
+#ifndef CONFIG_ZTEST
+    if (auto err = mCertHandler->UnsubscribeCertChanged(*this); !err.IsNone()) {
+        return Error(AOS_ERROR_WRAP(err), "failed to unsubscribe on cert changed event");
```
Comment (al1img) [inline src/smclient/smclient.cpp:651]: ditto

File: src/smclient/smclient.cpp [hunk 8]
```diff
@@ -618,6 +630,8 @@ Error SMClient::SetupChannel()
 
 Error SMClient::ReleaseChannel()
 {
+    mCertificateChanged = false;
```
Comment (al1img) [inline src/smclient/smclient.cpp:633]: Release channel is called unlocked. Move setting `mCertificateChanged` under lock.

**Review by MykolaSolyanko — APPROVED** [review]

File: src/iamclient/iamclient.cpp [hunk 9]
```diff
@@ -83,6 +83,12 @@ Error IAMClient::Stop()
     mClockSync->Unsubscribe(*this);
     mNodeInfoProvider->UnsubscribeNodeStatusChanged(*this);
 
+#ifndef CONFIG_ZTEST
```
Comment (mykola-kobets-epam) [inline src/iamclient/iamclient.cpp:86]: why ifndef required?

**Review by mykola-kobets-epam — APPROVED** [review]

File: src/iamclient/iamclient.cpp (hunk 9 above)
Comment (mlohvynenko) [reply src/iamclient/iamclient.cpp:86]: for mCertHandler :)
It's removed from members if CONFIG_ZTEST is defined

**Review by al1img — APPROVED** [review]

**Pull request merged**